molpipeline/mol2mol/filter.py

```diff
@@ -120,38 +164,207 @@ def pretransform_single(self, value: RDKitMol) -> OptionalMol:
         OptionalMol
             Molecule if it contains only allowed elements, else InvalidInstance.
         """
-        unique_elements = set(atom.GetAtomicNum() for atom in value.GetAtoms())
-        if not unique_elements.issubset(self.allowed_element_numbers):
-            forbidden_elements = unique_elements - self.allowed_element_numbers
+        to_process_value = (
+            Chem.AddHs(value) if 1 in self.allowed_element_numbers else value
```

Comment:
Implicitly modifying the molecule does not seem right. I understand that this can make sense when you are interested in exact element counts in the molecules. However, the filter should just filter molecules and not implicitly modify them before filtering. I think the most consistent way would be to let the user explicitly add hydrogens before the filter, which would result in an additional pipeline element. However, I think a compromise would also be fine by adding a flag `add_hs` to the `ElementFilter`'s constructor to let the user choose.

@c-w-feldmann what do you think?

Reply:
You're absolutely right, and I understand your point. I have 3 aspects I am unhappy with:

- It's not explicit but implicit, and I advocate the "Explicit is Better Than Implicit" principle. However, this might also be solved with good, explicit docstrings that you can add.
- The molecules might already have been prepared with Hydrogens added. Calling Chem.AddHs again adds additional computation for every molecule.
- I am not sure about our use cases, but for docking, for example, the users typically prepare the protonation states of their molecules with some other tool beforehand. In which case readding hydrogen might mess with the prepared molecules. Is something similar also possible in our use cases?

Reply:
Good point, could we check, whether hydrogens are there already? Or should we alternatively log a warning here if hydrogens are requested but the count is 0 and tell people to e.g. use an AddHs Pipeline module? (not sure whether this exists?)

Reply:
the three of us should talk about it together. This is a more conceptual decision we need to make.